### DISCUSSION TITLE
talking UART to a wifi module a la SoftwareSerial

### DISCUSSION THREAD

**Discussion by sornas** [Rahix/avr-hal#378] [Q&A] [answered]
hello

I've been trying to get my arduino uno to talk to an esp8266 wifi module, basically copying https://create.arduino.cc/projecthub/neverofftheinternet/esp8266-setup-and-first-wifi-connection-76fc3c. in the arduino.cc project they use SoftwareSerial which allows them to talk UART (USART?) with the esp over pins d2 and d3, and I'm trying to do the same.

there didn't seem to be a SoftwareSerial analog (hah) in this hal (which is reasonable tbf), so to that end I've looked at https://docs.rs/bitbang-hal/latest/bitbang_hal/serial/struct.Serial.html. there I need a [timer::CountDown](https://docs.rs/embedded-hal/latest/embedded_hal/timer/trait.CountDown.html) which looks a bit scary to just implement myself. so my question is if you know of anyone who has done something like that, or used the esp wifi in another way. otherwise I'll just go for it.

**Comment by Rahix** [comment, marked as answer]
I'm afraid that implementing `CountDown` is what you'll have to do here.  It shouldn't be too hard though, to set up one of the hardware timers for it.  You can maybe use the `HC SR-04` example as guidance on how to interact with a hardware timer: https://github.com/Rahix/avr-hal/blob/main/examples/arduino-uno/src/bin/uno-hc-sr04.rs

If you get stuck, let me know!

**Reply by sornas** [reply]
the example worked wonders, thanks! I think the timer works as it should, the problem now is that the serial implementation itself is buggy. if I apply a PR I can correctly send bytes to my computer, but if I echo back what I type on my keyboard it's mapped really weird.

for example, if i pressed the keys a-z in order I would expect this response:

```
00000000: 6162 6364 6566 6768 696a 6b6c 6d6e 6f70  abcdefghijklmnop
00000010: 7172 7374 7576 7778 797a 0a              qrstuvwxyz.
```

but what I get is:

```
00000000: 4323 6313 5333 730b 4b2b 6b1b 5b3b 7b07  C#c.S3s.K+k.[;{.
00000010: 4727 6717 5737 770f 4f2f 0a              G'g.W7w.O/.

note
lower nibble:  3333 333b bbbb bbb7 7777 777f ff
higher nibble: 4261 5370 4261 5370 4261 5370 42

and yes, it's the same mapping backwards.

00000000: 2f4f 0f77 3757 1767 2747 077b 3b5b 1b6b  /O.w7W.g'G.{;[.k
00000010: 2b4b 0b73 3353 1363 2343 0a              +K.s3S.c#C.
```

for now I think I can get away with running the esp on d0/d1. unless I can get my hands on a scope or something I don't know if I can debug the serial communication. if you (or someone else) want to try it out I created [a demo repository](https://github.com/sornas/arduino-uno-uart).
